### ISSUE TITLE
How to install Microsoft Visual C++ 2019 redistributables

### ISSUE THREAD

**Issue by mariusbancila** [dotnetinstaller/dotnetinstaller#140]
I have a setup for a VC++ application built with dotnetinstaller. Being a native application, the VC++ redistributables are a prerequisite. However, nowadays these change very often (every 6-8 weeks). This means we release a new version of the application, but newer version of the redistributables are also released and installed on a customer machine before the customer actually installs our application. The result is the old redistributables from our setup will fail to install, since a newer version exists. This, in turn, results in the failure of the setup, therefore the application MSI will not be executed.

Currently, the workaround is to open a command prompt elevated as administrator and run the msiexec with the right parameters. I want to find a better experience for the customers.

The VC++ redistributables are added as required prerequisites. That means the `required_install` property is set to true. The result of this is that failure to install the prerequisite means installation failure. Although `allow_continue_on_error` and `default_continue_on_error` are set to true, the user is informed about the failure and can continue with other prerequisites, but the setup fails in the end and the MSI execution does not start.

I have tried making the VC++ redistributable optional prerequisites by setting the `required_install` property to false. However, this leads to another undesired behavior. If any required prerequisite is missing, then the dialog of prerequisites is displayed and I can check/uncheck the optional ones (and therefore skip them if I don't want them). However, if all required prerequisites are found and only the optional ones are missing, the setup does not display the prerequisites dialog and go straight to execute the MSI.

I need an option to make these pre-requisites optional in the sense that failure to install them will not stop the execution of the MSI; however, I do want the users to see they are missing, if the case, and have the option to install them.

Is this possible in some way with dotnetintaller?

**Comment by dblock** [comment]
I wouldn't ignore errors, it could be a legit one. I think you want an install check that skips installation of the VC++ redistributable when a newer or same version is already installed. This way it just skips installation but the pre-requisite is satisfied. Like [here](https://github.com/dotnetinstaller/dotnetinstaller/blob/master/InstallerEditor/wizardcomponents.xml#L9).

**Comment by mariusbancila** [comment]
I understand the point, and I agree, but I could not find a way to do it. In your example, the `installedcheck` looks like this:
```
<installedcheck path="SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{7299052b-02a4-4627-81f2-1818da5d550d}" fieldname="Version" fieldvalue="1" defaultvalue="False" fieldtype="REG_DWORD" comparison="exists" rootkey="HKEY_LOCAL_MACHINE" wowoption="NONE" type="check_registry_value" description="Installed Check" />
```
In my case, it looks like this:
```
<installedcheck path="SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{855e31d2-9031-46e1-b06d-c9d7777deefb}" fieldname="" fieldvalue="" defaultvalue="False" fieldtype="REG_SZ" comparison="exists" rootkey="HKEY_LOCAL_MACHINE" wowoption="WOW64_32" type="check_registry_value" description="Installed Check" />
```
I am checking if the key exists, not if a name-value pair has some value. This is how the Registry looks for VC++ 2019 redistributables.
![image](https://user-images.githubusercontent.com/5336650/114608470-65b81880-9ca6-11eb-949b-068d96248f83.png)
The `DisplayName` stores the version, but this is a string, not an integer. The latest version is 14.28.29913. The next one may be 14.28.30123 (just a number) or 14.29.30123, or maybe 15.xx.xxx something.

**Comment by icnocop** [comment]
Instead of using `comparison="exists"` it would be best to use `comparison="version_ge"` which "Check[s] whether the version in registry is greater or equal than the check value.".

See the definition of `version_ge` [here](https://github.com/dotnetinstaller/dotnetinstaller/blob/36567509faef246ce95af6d567c945ff0bb90a22/Documentation/Content/InstalledChecks.aml#L142-L146):

```
<definedTerm>version_ge</definedTerm>
<definition>
The registry value is interpreted as a file version. Check whether the version in registry is
greater or equal than the check value.
</definition>
```

An example can be found [here](https://github.com/dotnetinstaller/dotnetinstaller/blob/ed51319f6e27435f4ec3ceb19ea6b6f61b3ad2f4/InstallerEditor/wizardcomponents.xml#L118):

```
<installedcheck path="SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Client" fieldname="Version" fieldvalue="4.0.30319" defaultvalue="False" fieldtype="REG_SZ" comparison="version_ge" rootkey="HKEY_LOCAL_MACHINE" wowoption="NONE" type="check_registry_value" description="Installed Check" />
```

**Comment by mariusbancila** [comment]
Does not work with VC++ redistributables, because each version has a new product code and implicitly a new path under `SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\`.

**Comment by dblock** [comment]
But isn't there a common key somewhere for the installed vc redistributable?

**Comment by mariusbancila** [comment]
Yes, there seems to be the `HKEY_LOCAL_MACHINE\SOFTWARE[\Wow6432Node]\Microsoft\VisualStudio\14.0\VC\Runtimes\{x86|x64|ARM}` key. However, it's unfortunate how they store the version:
![image](https://user-images.githubusercontent.com/5336650/114776742-7ab2ab80-9d7b-11eb-9e3b-a1954f8a14d0.png)
It has a leading `v` which results in the following error:
![image](https://user-images.githubusercontent.com/5336650/114776801-8aca8b00-9d7b-11eb-8ab7-6c017aff27e8.png)

**Comment by dblock** [comment]
Bah. Comparing major/minor/build is next, but I can see how this is annoying.

I think you should extend https://github.com/dotnetinstaller/dotnetinstaller/blob/36567509faef246ce95af6d567c945ff0bb90a22/dotNetInstallerToolsLib/FileUtil.cpp#L220 to support `v[version]`.

**Comment by icnocop** [comment]
You can add more than one `installedcheck` and use an `installedcheckoperator`.

An example can be found [here](https://github.com/dotnetinstaller/dotnetinstaller/blob/36567509faef246ce95af6d567c945ff0bb90a22/Samples/InstallCheckOperators/Configuration.xml#L14-L20).

```
      <installedcheckoperator type="And" description="Install if .NET 1.1 SP1 IS NOT installed">
        <installedcheck path="SOFTWARE\Microsoft\NET Framework Setup\NDP\v1.1.4322" fieldname="Install" fieldvalue="1" defaultvalue="False" fieldtype="REG_DWORD" comparison="match" rootkey="HKEY_LOCAL_MACHINE" wowoption="NONE" type="check_registry_value" description="Installed" />
        <installedcheckoperator type="Or" description="SP1 or SP2">
          <installedcheck path="SOFTWARE\Microsoft\NET Framework Setup\NDP\v1.1.4322" fieldname="SP" fieldvalue="1" defaultvalue="False" fieldtype="REG_DWORD" comparison="match" rootkey="HKEY_LOCAL_MACHINE" wowoption="NONE" type="check_registry_value" description="SP1" />
          <installedcheck path="SOFTWARE\Microsoft\NET Framework Setup\NDP\v1.1.4322" fieldname="SP" fieldvalue="2" defaultvalue="False" fieldtype="REG_DWORD" comparison="match" rootkey="HKEY_LOCAL_MACHINE" wowoption="NONE" type="check_registry_value" description="SP2" />
        </installedcheckoperator>
      </installedcheckoperator>
```

Documentation can be found [here](https://github.com/dotnetinstaller/dotnetinstaller/blob/36567509faef246ce95af6d567c945ff0bb90a22/Documentation/Content/InstalledChecks.aml#L233-L260).

>           You can combine results from multiple elements by adding one or more <newTerm>InstalledCheck</newTerm> elements under 
>           an <newTerm>InstalledCheckOperator</newTerm>. You may add operators under other operators to achieve more complex logic.
>           The following operators are supported: <newTerm>And</newTerm>, <newTerm>Or</newTerm> and <newTerm>Not</newTerm>.

So basically I would check and compare the `Installed`, `Major`, `Minor`, `Bld`, and `Rbld` registry values.

**Comment by icnocop** [comment]
Support for v[version] format for version numbers was implemented in https://github.com/dotnetinstaller/dotnetinstaller/pull/141

**Comment by icnocop** [comment]
Implemented in [dotNetInstaller v3.1.415](https://github.com/dotnetinstaller/dotnetinstaller/releases/tag/3.1.415)
